tests/dashboard/test_product.py

```diff
@@ -255,6 +261,56 @@ def test_view_product_edit(admin_client, product):
         "dashboard:product-details", kwargs={"pk": product.pk}
     )
     assert product.name == "Product second name"
+    assert (
+        mocked_set_tax_rate_code.call_count == 0
+    ), "The tax code shouldn't have been changed"
+
+
+def test_update_product_tax_rate(admin_client, product):
+    """Ensure tax rates are correctly listed in the form and the form is properly
+    updating the tax rate.
+    """
+    tax_rate_code = "dummy"
+
+    url = reverse("dashboard:product-update", kwargs={"pk": product.pk})
+    data = {
+        "name": "Product second name",
+        "description": "Product description.",
+        "price_0": 10,
+        "price_1": product.price.currency,
+        "category": product.category.pk,
+        "variant-sku": "123",
+        "variant-quantity": 10,
+        "tax_rate": tax_rate_code,
+    }
+
+    manager = get_extensions_manager()
+
+    with mock.patch("saleor.extensions.manager.ExtensionsManager") as mocked_manager:
+        mocked_manager = mocked_manager()
+        mocked_manager.assign_tax_code_to_object_meta.side_effect = (
+            manager.assign_tax_code_to_object_meta
+        )
+        mocked_manager.get_tax_rate_type_choices.return_value = [
+            TaxType(code=tax_rate_code, description="")
+        ]
+
+        # Ensure the available tax rate codes are properly returned
+        response = admin_client.get(url)
```

Comment:
You should have two separate tests here. You are testing two different things. 1) retrieving tax data from the form. 2) updating tax data for product